### PR TITLE
fix(generator): streaming writes with routing instructions

### PR DESCRIPTION
**Pull request by coryan**
The generator cannot create code to capture the routing instructions on
asynchronous streaming writes.  The generated functions do not have a
request to extract the routing instructions from.  Just inject the
basic metadata, the caller is responsible for adding the routing
instructions.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9524)

### CODE REVIEW

File: generator/internal/metadata_decorator_generator.cc [hunk 1]
```diff
@@ -397,16 +397,21 @@ std::unique_ptr<::google::cloud::internal::AsyncStreamingReadRpc<
...
           R"""(
 std::unique_ptr<::google::cloud::internal::AsyncStreamingWriteRpc<
     $request_type$, $response_type$>>
 $metadata_class_name$::Async$method_name$(
     google::cloud::CompletionQueue const& cq,
     std::unique_ptr<grpc::ClientContext> context) {
-)""",
-          SetMetadataText(method, kPointer),
-          R"""(
+  SetMetadata(*context);
```
Comment (dbolduc) [inline generator/internal/metadata_decorator_generator.cc:414]: hm, so the comment you wrote also applies to async streaming read-writes. It seems like we should either copy the comment, or have both use `SetMetadataText(method, kPointer)`. Then in that method we would say like:
```cc
std::string SetMetadataText(...) {
  std::string const context = ...;

  // blah blah blah
  // comment about client streaming
  // blah blah blah
  if (method.client_streaming()) return "  SetMetadata(" + context + ");";
  ...
}
```

File: generator/internal/metadata_decorator_generator.cc [hunk 2]
```diff
@@ -397,16 +397,21 @@ std::unique_ptr<::google::cloud::internal::AsyncStreamingReadRpc<
       continue;
     }
     if (IsStreamingWrite(method)) {
+      // Asynchronous streaming writes do not consume a request, typically the
```
Comment (dbolduc) [inline generator/internal/metadata_decorator_generator.cc:400]: s/, typically/. Typically/

File: generator/internal/metadata_decorator_generator.cc [hunk 3]
```diff
@@ -397,16 +397,21 @@ std::unique_ptr<::google::cloud::internal::AsyncStreamingReadRpc<
...
+      // first `Write()` call contains any relevant data to "start" the stream.
+      // Thus, the decorator cannot add any routing instructions. The caller
+      // should initialize `context` with any such instructions. At this time,
```
Comment (dbolduc) [inline generator/internal/metadata_decorator_generator.cc:403]: I think I would just say: "The handwritten caller should initialize \`context\` with any such instructions." and delete the rest of the paragraph.

If you decide to keep the rest of the paragraph then, `s/, ut/, but/` a few lines down.

File: generator/internal/metadata_decorator_generator.cc (hunk 1 above)
Comment (coryan) [reply generator/internal/metadata_decorator_generator.cc:415]: I chose to copy the comment

File: generator/internal/metadata_decorator_generator.cc (hunk 2 above)
Comment (coryan) [reply generator/internal/metadata_decorator_generator.cc]: Fixed.

File: generator/internal/metadata_decorator_generator.cc (hunk 3 above)
Comment (coryan) [reply generator/internal/metadata_decorator_generator.cc]: Option 1

**Review by dbolduc — APPROVED** [review]

**Pull request merged**